scanner/services/indexer.go

```diff
@@ -0,0 +1,140 @@
+package services
+
+import (
+	"context"
+	"crypto/sha512"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/quay/claircore"
+	"github.com/quay/zlog"
+	v4 "github.com/stackrox/rox/generated/internalapi/scanner/v4"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/grpc/authz/allow"
+	"github.com/stackrox/rox/scanner/indexer"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type indexerService struct {
+	v4.UnimplementedIndexerServer
+	indexer indexer.Indexer
+}
+
+// NewIndexerService creates a new indexer service.
+func NewIndexerService(indexer indexer.Indexer) *indexerService {
+	return &indexerService{
+		indexer: indexer,
+	}
+}
+
+func (s *indexerService) CreateIndexReport(ctx context.Context, req *v4.CreateIndexReportRequest) (*v4.IndexReport, error) {
+	// TODO We currently only support container images, hence we assume the resource
+	//      is of that type. When introducing nodes and other resources, this should
+	//      evolve.
+	resourceType := "containerimage"
+	if err := validateContainerImageRequest(req); err != nil {
+		return nil, err
+	}
+	// Create a claircore.Digest using the full Hash ID, including resource type.
+	hashIDSum := sha512.Sum512([]byte(req.GetHashId()))
+	manifestDigest, err := claircore.NewDigest(claircore.SHA512, hashIDSum[:])
+	if err != nil {
+		return nil, fmt.Errorf("internal error: creating container image manifest digest: %w", err)
+	}
+
+	ctx = zlog.ContextWithValues(ctx,
+		"component", "scanner/service/indexer",
+		"resource_type", resourceType,
+		"manifest_digest", manifestDigest.String())
+
+	// Setup authentication.
+	var opts []indexer.Option
+	hasAuth := req.GetContainerImage().GetUsername() != ""
```

Comment:
what happens if we just blindly set `authn.Basic` to `""` username and password? Will this not work?

Reply:
Not sure what you mean. Please, specifically clarify "blidly set" and "work".

Reply:
It's possible that the default values could be considered invalid options. Or maybe the implementation will send the authorization headers with blank values. Either way, I think it doesn't really matter because we have the choice to include it or not and still ensure safety.